shared/rollouts/__init__.py

```diff
@@ -106,45 +114,83 @@ def _buckets(self):
         this class will compute a hash including:
         - The experiment's name
         - The repo's id
-        - (Optional) A unique salt
+        - A salt
 
         The range of possible hash values is divvied up into buckets based on the
         `proportion` and `variants` members. The hash for this repo will fall into
         one of those buckets and the corresponding variant (or default value) will
         be returned.
         """
-        test_population = int(self.proportion * Feature.HASHSPACE)
+        test_population = int(self.feature_flag.proportion * Feature.HASHSPACE)
 
         buckets = []
         quantile = 0
-        for variant in self.variants.values():
+        for variant in self.ff_variants:
             quantile += variant.proportion
             buckets.append((int(quantile * test_population), variant))
 
         return buckets
 
-    # NOTE: `@lru_cache` on instance methods is ordinarily a bad idea:
+    def _get_override_variant(self, identifier):
+        """
+        Retrieves the FeatureFlagVariant applicable to the given identifer according to
+        defined overrides. Returns None if no override is found.
+        """
+        for variant in self.ff_variants:
+            if (
+                identifier in variant.override_owner_ids
+                or identifier in variant.override_repo_ids
+            ):
+                return variant
+        return None
+
+    def _is_valid_rollout(self):
```

Comment:
maybe we run this in the function that fetches values from the DB, and if it isn't valid log a warning. that way somebody making a mistake in the web UI (or the cache clearing while they're still in the middle of making updates) doesn't totally shut down every feature

we use the python standard library logger a la https://github.com/codecov/worker/blob/main/helpers/checkpoint_logger/__init__.py#L22